src/Servers/Kestrel/Core/src/Internal/Http3/Http3ControlStream.cs

```diff
@@ -0,0 +1,394 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.IO.Pipelines;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Connections.Abstractions.Features;
+using Microsoft.AspNetCore.Hosting.Server;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http3.QPack;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http3
+{
+    internal abstract class Http3ControlStream : IHttp3Stream, IThreadPoolWorkItem
+    {
+        private const int ControlStream = 0;
+        private const int EncoderStream = 2;
+        private const int DecoderStream = 3;
+        private Http3FrameWriter _frameWriter;
+        private readonly Http3Connection _http3Connection;
+        private HttpConnectionContext _context;
+        private readonly Http3Frame _incomingFrame = new Http3Frame();
+        private volatile int _isClosed;
+        private int _gracefulCloseInitiator;
+
+        private bool _haveReceivedSettingsFrame;
+
+        public Http3ControlStream(Http3Connection http3Connection, HttpConnectionContext context)
+        {
+            var httpLimits = context.ServiceContext.ServerOptions.Limits;
+
+            _http3Connection = http3Connection;
+            _context = context;
+
+            _frameWriter = new Http3FrameWriter(
+                context.Transport.Output,
+                context.ConnectionContext,
+                this,
+                context.TimeoutControl,
+                httpLimits.MinResponseDataRate,
+                context.ConnectionId,
+                context.MemoryPool,
+                context.ServiceContext.Log);
+
+            var closedRegistration = _context.ConnectionContext.ConnectionClosed.Register(state => ((Http3ControlStream)state).OnStreamClosed(), this);
```

Comment:
Why is this here

Reply:
If a control stream is closed, abort the entire connection.

Reply:
This needs to be disposed